File: app/src/main/java/com/pr0gramm/app/ui/intro/slides/ThemeActionItemsSlide.java
package com.pr0gramm.app.ui.intro.slides;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.View;

import com.google.common.collect.FluentIterable;
import com.pr0gramm.app.Settings;
import com.pr0gramm.app.services.ThemeHelper;
import com.pr0gramm.app.ui.Themes;

import java.util.List;

/**
 */
public class ThemeActionItemsSlide extends ActionItemsSlide {
    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        view.setBackgroundResource(ThemeHelper.theme().accentColor);
    }

    @Override
    protected String getIntroTitle() {
        return "Farbe";
    }

    @Override
    protected String getIntroDescription() {
        return "Bist du Neuschwuchtel oder erinnerst du Dich noch an den Kampf der Farben? Wähle deinen persönlichen Sieger.";
    }

    @Override
    protected List<ActionItem> getIntroActionItems() {
        return FluentIterable.from(Themes.values())
                .transform(ThemeActionItem::new)
                .filter(ActionItem.class)
                .toList();
    }

    @Override
    protected boolean singleChoice() {
        return true;
    }

    private class ThemeActionItem extends ActionItem {
        final Settings settings = Settings.of(getContext());
        final Themes theme;

        ThemeActionItem(Themes theme) {
            super(theme.title(getContext()));
            this.theme = theme;
        }

        @Override
        public boolean enabled() {
            return settings.themeName().equals(theme.name());
        }

        @Override
        public void activate() {
            ThemeHelper.updateTheme(getContext(), theme);

            View view = getView();
            if (view != null) {
                view.setBackgroundResource(theme.primaryColor);
            }
        }

        @Override
        public void deactivate() {
        }
    }
}

File: app/src/main/java/com/pr0gramm/app/ui/intro/slides/ThemeActionItemsSlide.kt
package com.pr0gramm.app.ui.intro.slides

import com.pr0gramm.app.Settings
import com.pr0gramm.app.services.ThemeHelper
import com.pr0gramm.app.ui.Themes

/**
 */
class ThemeActionItemsSlide : ActionItemsSlide("ThemeActionItemsSlide") {
    override val introTitle: String = "Farbe"

    override val introDescription: String =
            "Bist du Neuschwuchtel oder erinnerst du Dich noch an den Kampf der Farben? Wähle deinen persönlichen Sieger."

    override val introActionItems: List<ActionItem>
        get() = Themes.values().map { ThemeActionItem(it) }

    override val singleChoice: Boolean = true

    override val introBackgroundResource: Int = ThemeHelper.theme.accentColor

    private inner class ThemeActionItem(val theme: Themes) : ActionItem(theme.title(context)) {
        private val settings = Settings.get()

        override fun enabled(): Boolean {
            return settings.themeName == theme.name
        }

        override fun activate() {
            ThemeHelper.updateTheme(theme)
            view?.setBackgroundResource(theme.primaryColor)
        }

        override fun deactivate() {}
    }
}
